### PR TITLE
Reorder data for RNN::Train() and other RNN functions

### PR DESCRIPTION
**Pull request by rcurtin**
This is related to a discussion from long ago, but I am not sure I am finding the right ticket to reference: #513, #831, #541?  Anyway, if I am remembering right, there was some discussion about whether to use `mat` or `cube` to represent input to neural networks.  It's completely possible my views have changed over time and I don't realize the change, but to me now it seems the best way is to have `mat` be the input to FFNs (i.e. one column per point, one row per dimension), but `cube` be the input to RNNs (i.e. one column per point, one row per dimension, one slice per timestep).

This PR implements that support and adds some documentation to the `RNN` class about how the data is expected to be formatted.  There are also a couple of other warning fixes that I noticed and handled.

One of the important reasons for doing this change is that internally, at each time step, the layer expects a matrix of shape (n_dims x n_points).  But RNN data is currently embedded in a matrix of shape ((n_dims * n_timesteps) * n_points), with a single column containing all of the time steps for a given point.  This means that when the batch size is greater than 1 (which it typically will be after #1137) we have to extract a non-contiguous submatrix, which Armadillo will do as a copy.  If we reorder our data into a cube where each timestep is a separate slice, this allows us to avoid that copy.  I also think the cube structure is a easier to understand for a user, but we can change that around if there are other opinions.

This doesn't solve the problem of how to pass sequences of different length to RNNs---you still must either pass them individually, or zero-pad them all into a cube and pass that.  That problem can be solved some other day...

### CODE REVIEW

File: src/mlpack/core/math/shuffle_data.hpp [hunk 1]
```diff
@@ -46,11 +47,13 @@ void ShuffleData(const MatType& inputPoints,
...
                  MatType& outputPoints,
                  LabelsType& outputLabels,
-                 const std::enable_if_t<arma::is_SpMat<MatType>::value>* = 0)
+                 const std::enable_if_t<arma::is_SpMat<MatType>::value>* = 0,
+                 const std::enable_if_t<!arma::is_Cube<MatType>::value>* = 0)
 {
   // Generate ordering.
   arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
       inputPoints.n_cols - 1, inputPoints.n_cols));
+//  std::cout << "ordering:\n" << ordering.t();
```
Comment (zoq) [inline src/mlpack/core/math/shuffle_data.hpp]: Looks like we can remove the debug message here.

File: src/mlpack/core/math/shuffle_data.hpp [hunk 2]
```diff
@@ -67,11 +70,80 @@ void ShuffleData(const MatType& inputPoints,
...
-  outputLabels.set_size(inputLabels.n_elem);
-  for (size_t i = 0; i < inputLabels.n_elem; ++i)
-    outputLabels[ordering[i]] = inputLabels[i];
+  if (&inputPoints == &outputPoints || &inputLabels == &outputLabels)
+  {
+    MatType newOutputPoints(locations, values, inputPoints.n_rows,
+        inputPoints.n_cols, true);
+    LabelsType newOutputLabels(inputLabels.n_elem);
+    for (size_t i = 0; i < inputLabels.n_elem; ++i)
+      newOutputLabels[ordering[i]] = inputLabels[i];
```
Comment (zoq) [inline src/mlpack/core/math/shuffle_data.hpp]: Can you use a non-contiguous view here, to avoid the for loop? It's probably the same.

File: src/mlpack/tests/math_test.cpp [hunk 3]
```diff
@@ -631,6 +631,8 @@ BOOST_AUTO_TEST_CASE(SparseShuffleTest)
     data(0, i) = i;
     labels[i] = i;
   }
+  // This appears to be a necessary workaround for an Armadillo 8 bug.
+  data *= 1.0;
```
Comment (zoq) [inline src/mlpack/tests/math_test.cpp]: Interesting, perhaps we should update the matrix build, maybe we already test against armadillo 8?

File: src/mlpack/tests/recurrent_network_test.cpp [hunk 4]
```diff
@@ -690,13 +698,23 @@ void DistractedSequenceRecallTestNetwork(
...
       model.Predict(input, output);
-      data::Binarize(output, output, 0.5);
+      for (size_t j = 0; j < output.n_slices; ++j)
+      {
+        arma::mat outputSlice = output.slice(j);
+        data::Binarize(outputSlice, outputSlice, 0.5);
+        output.slice(j) = outputSlice;
+      }
+
+//      std::cout << "label:\n" << label << "\noutput:" << output
```
Comment (zoq) [inline src/mlpack/tests/recurrent_network_test.cpp]: Looks like another debug output.

File: src/mlpack/tests/recurrent_network_test.cpp [hunk 5]
```diff
@@ -690,13 +698,23 @@ void DistractedSequenceRecallTestNetwork(
...
+      arma::cube output;
+      arma::cube input(testInput.at(0, i).memptr(), inputSize, 1,
+          testInput.at(0, i).n_elem / inputSize, false, true);
 
       model.Predict(input, output);
-      data::Binarize(output, output, 0.5);
+      for (size_t j = 0; j < output.n_slices; ++j)
+      {
+        arma::mat outputSlice = output.slice(j);
+        data::Binarize(outputSlice, outputSlice, 0.5);
```
Comment (zoq) [inline src/mlpack/tests/recurrent_network_test.cpp]: Do you think we should add ``cube`` support, might be helpful to postprocess the data. If you agree, this could be a good beginner task.

File: src/mlpack/tests/math_test.cpp (hunk 3 above)
Comment (rcurtin) [reply src/mlpack/tests/math_test.cpp]: I'm in the process of updating the matrix build, but it's not ready yet.  I haven't isolated the Armadillo bug yet, but I'll either open an issue or send a fix to Conrad.

File: src/mlpack/tests/recurrent_network_test.cpp (hunk 5 above)
Comment (rcurtin) [reply src/mlpack/tests/recurrent_network_test.cpp]: Yeah, I think it would be a nice beginner task to add.  I'll go ahead and write up an issue for it.  I think no new code is necessary; just the existing `Binarize()` can be adapted for this case, then this code can be simplified.

File: src/mlpack/core/math/shuffle_data.hpp (hunk 2 above)
Comment (rcurtin) [reply src/mlpack/core/math/shuffle_data.hpp]: Ah, I didn't think of that---fixed.

File: src/mlpack/core/math/shuffle_data.hpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/core/math/shuffle_data.hpp]: Oops, fixed.
